### ISSUE TITLE
Determine how to publish Open Project specifications

### ISSUE THREAD

**Issue by jamsden** [oslc-op/oslc-admin#19]
Contains a sample OSLC OP Project Specification that follows these guidelines and is what would be proposed to be published in the OASIS Document Library as PS01.

Content for OASIS standards-track documents require the authoritative, normative specification document to be fixed, immutable, and at a location that OASIS can control or at least can vouch for.

OASIS TC Committee specifications are published in the OASIS library: https://docs.oasis-open.org.  Only OASIS staff have access to docs.oasis-open.org.  

Anything published to docs.oasis-open.org must follow OASIS naming guidelines. For multi-part specifications:
```
https://docs.oasis-open-project.org/[wgShortName]/[shortName]/v[oslcVersion]/[specStatus][revision]/[partNumber]-[partName]/[shortName]-v[oslcVersion]-[specStatus][revision]-[partNumber]-[partName].[ext]
```

The OSLC OP uses GitHub as its document repository. GitHub meets the requirements that the authoritative, normative specification document to be fixed, immutable location. But it has two problems. First the GitHub repo is not under just OASIS control, the oslc-op TCS also has admin access. And second, GitHub does not do content negotiation and does not render HTML. So any Project Specification that requires rendering needs special handling.

Although it is possible to use raw.githack.org and version ReSpec with the revision of a specification, and put the URL to the ReSpec revision in the spec, it adds complexity, risk and slows performance to have to render the HTML every time it's accessed. 

If we think of ReSpec as being a document processor, and consider a build as something that creates the HTML to be published, then this fits a typical SCM product delivery model. In these models, the authoritative sources is immutable, and the results of the build is distributed to consumers. 

Typical build parameters for OP specification publication could include:
* specStatus - PSD, PS, COS, OS (defaults to PSD)
* publishDate - defaults to the date the build executes
* revision
ReSpec can set all these parameters as URI query parameters that override whatever is in the file.

## Proposed OP Publishing Process

For OASIS Project Specifications that require some document rendering process (such as ReSpec), the authoritative, normative specification, fixed and immutable Project Specification will be the tagged source document in the GitHub repository.

An automated build process will be used to create a published version of the document by rendering the source in static HTML and PDF formats. The HTML document will be marked as Authoritative, meaning the source it was generated from is actually normative. The build process is automated to ensure no changes are introduced between the source and rendered document.

The rendered HTML and PDF files for Project Specifications, Candidate OASIS Standards, and OASIS Standards are stored in the OASIS OP document library.

The URL to the authoritative version of the published specification is the URL of the tagged version in the repo. Tag names use the following naming convention: `[shortName]-v[oslcVersion]-[specStatus][revision]`. OASIS Specifications do not have the ['revision] component.

Project Specifications stored in the OASIS OP document library use the following naming conventions (both single and multi-part specifications use the same naming convention):
```
https://docs.oasis-open-project.org/[wgShortName]/[shortName]/v[oslcVersion]/[specStatus][revision]/[file-name].[ext]
```
And the latestVersion URI will redirect to the latest published specification:
```
https://docs.oasis-open.org/[wgShortName]/[shortName]/v[oslcVersion]/[file-name].[ext]
```

For example, The OSLC Change Management specification is a multi-part specification with two parts, one for the specification and another for the vocabulary. A specific Project Specification Draft revision would have something like the following:
```
This version:
https://raw.githubusercontent.com/oslc-op/oslc-specs/cm-v3.0-psd01/specs/cm/change-mgt-spec.html (Authoritative Source)
https://docs.oasis-open.org/oslc-op/cm/v3.0/psd01/change-mgt-spec.html 
https://docs.oasis-open.org/oslc-op/cm/v3.0/psd01/change-mgt-spec.pdf

Previous version:
http://docs.oasis-open.org/oslc-domains/cm/v3.0/cs02/part1-change-mgt/cm-v3.0-cs02-part1-change-mgt.html
http://docs.oasis-open.org/oslc-domains/cm/v3.0/cs02/part1-change-mgt/cm-v3.0-cs02-part1-change-mgt.pdf

Latest version:
https://docs.oasis-open.org/oslc-op/cm/v3.0/change-mgt-spec.html
https://docs.oasis-open.org/oslc-op/cm/v3.0/change-mgt-spec.pdf
```
We agree that Editor's draft URLs will not be include in project specifications until we have a means of establishing stable URLs for them, similar to the Latest version URLs maintained by OASIS. This is because GitHub URLs are not necessarily stable. The URLs go through redirects that include implementation components that might change.

## Potential Issues

- [x] OASIS requires the documents to have a Change History entry what includes the revision number. 

This would have to be added before the builder assigns the revision. OASIS has indicated that the Change History section is optional for OP specifications. We could consider retaining it and deal with the minor editing issue regarding the table including a future revision number. Alternatively we can provide this information in the tag message.

- [x] Following the OASIS naming guidelines breaks relative links across multi-part specifications because the version, revision, status and part are added to the file name.

Follow the OASIS naming guidelines for the path, but not the file. The reason is to keep all the variability in the path, not in the file name. If we need part numbers in the file, we can do that by renaming the file and editing all the relative links. But the revision and specStatus can't be in the file name without breaking relative links. 

OASIS has agreed to apply the naming conventions to only the path components, not the file for OP specifications. This preserves relative link references across multi-part specifications.

- [x] Git tags (or releases) apply to the repo, not individual files or folders in the repo. This will require all the documents in https://github.com/oslc-op/oslc-specs to be released at the same time.

This can be addressed by following the OASIS versioning naming conventions: `[shortName]/v[oslcVersion]/[specStatus][revision]`, replacing the / with -. For example, Change Management Project Specification revision 01 would be tagged `cm-v3.0-ps01`. Although everything in the repo is tagged with this tag, users would understand that it only applies to change management specifications.  

- [ ] Git allows tags to be removed and changed. 

This will need to be restricted somehow to ensure tagged versions are never changed.

**Comment by berezovskyi** [comment]
I think you meant not to put a slash between the status and the revision: `https://docs.oasis-open.org/oslc-op/cm/v3.0/psd03/change-mg-vocab.html`

I would also go further and suggest to combine the spec name and version as OASIS considers another version to be a completely new spec: `https://docs.oasis-open.org/oslc-op/cm-3.0/psd03/change-mg-vocab.html`

Overall, I like the proposal. We can then host WDs on open-services or in GH pages.

**Comment by berezovskyi** [comment]
P.S. I just spent 15 minutes trying to figure out a nifty way to trigger a respec build only if a file in the spec dir was edited (again, to prevent people from having false hopes of very active spec editing). I did and want to leave this here for the future reference:

`git show --name-only --oneline | tail -n +2 | grep specs/cm/ | wc -l` will return 0 if no files were changed in the CM spec in the last commit or a number of changed files otherwise.

One could use this result with the `-ge 0` check in Bash to decide whether to trigger the build.

**Comment by berezovskyi** [comment]
And the manual for the CI integration with GH pages: https://circleci.com/blog/deploying-documentation-to-github-pages-with-continuous-integration/

**Comment by berezovskyi** [comment]
@ndjc how do you run ReSpec from the command line to bake static HTML files? The tool described in https://github.com/w3c/respec/wiki/ReSpec-Editor's-Guide#creating-a-static-snapshot apparently works only with very recent versions of ReSpec.

**Comment by berezovskyi** [comment]
@ndjc I really hope this is a joke and there is a more automated way to do this: https://github.com/shexSpec/spec/blob/gh-pages/publishing-process.md

**Comment by ndjc** [comment]
I do not believe our version of ReSpec has a way to run it from the command line. I could of course add such a capability one day, but not this week! Another argument in favor of junking our difficult-to-maintain branch and go back to using the W3C version - though that would not be at all cheap since we would still need to keep many of the OASIS variations.

And as for the overall process described by the shex md file, that sounds about right, except the initial copy would be just creating a branch in git, and we do not currently have (or need?) the HTML diff steps.

**Comment by berezovskyi** [comment]
I think I finally cracked this (and the publication process). To fix it in the best way possible would require a breaking change, but the effort is relatively minor.

So, I almost gave up on editing the source to produce the right dates and was about to go all in with a CI pipeline when I found out that ReSpec does not have a good CLI. Not until version 20.10, at least.

But ShEx somehow showed a reasonable draft modification date (2 weeks ago, matches the commit) without stating it explicitly. From the knowledge @ndjc equipped me earlier, I started investigating the `document.lastModified` property returned by Github Pages and GitHack. Turns out GH Pages return a truthful value, while GitHack always returns `now()`.

The obvious next step was to figure out to switch to GH Pages. I then realised that ShEx pushes straight to the `gh-pages` branch. I branched off our `master` and pushed and I saw the spec and the modification date corresponded to the time of push! 🎉 However, there were 2 problems: `gh-pages` is an ugly branch and the timestamps for _all_ specs get updated at the same time. For the former, I just configured the repo to publish the pages straight from the master branch.

But the other problem is not easily solvable:

1. we can still have a CI script that will deploy the `/specs` folder on e.g. `specs.open-services.net` or
2. we split all specs into individual repos and then the update to one spec will not disturb the last modification date of the other specs (I prefer this approach, plus version tags and issues will be grouped by the spec; we can let OASIS publish the PSDs).
3. Or option 2 for the WDs and option 1 for the PSDs.

I just updated the dates and statuses on the RM spec and tomorrow you will see that the modification for it is still July 1st: https://oslc-op.github.io/oslc-specs/specs/rm/requirements-management-spec.html

We went full circle to GH Pages as @jamsden was initially suggesting 🤸‍♂ 

And yay, @axelreichwein, no more blocking of those hacker git whatever sites 😄
